### PR TITLE
Clear Architecture field in platform constraint for arm architectures

### PR DESCRIPTION
**Pull request by nishanttotla**
This is a potential fix for https://github.com/docker/swarmkit/issues/2294

This is a temporary requirement.

Also related: https://github.com/opencontainers/image-spec/issues/661

### CODE REVIEW

**Comment by aaronlehmann** [comment]
I wonder if we could append all known ARM variants to `platforms` in this case.

**Comment by nishanttotla** [comment]
Ping @stevvooe

**Comment by nishanttotla** [comment]
@aaronlehmann based on the discussion in the SwarmKit issue, I don't think we should be doing that. This will eventually be fixed for images.

**Comment by aaronlehmann** [comment]
Wouldn't this PR cause ARM images to be scheduled on x86 and other platforms? I suppose it's better than not working at all, but I'm wondering if we can do better.

**Comment by nishanttotla** [comment]
@aaronlehmann yes, that will happen. I think until the manifest is fixed, it is better to err on the side of putting lesser constraints.

**Comment by stevvooe** [comment]
This fix needs to happen server-side:

1. If the platform comes from a manifest list or oci index, it is fine.
2. If it has to be resolved from the image config, it should undergo similar logic.

> Wouldn't this PR cause ARM images to be scheduled on x86 and other platforms? I suppose it's better than not working at all, but I'm wondering if we can do better.

Yes, but right now, we have no way of resolving which version of arm an image belongs to.

**Comment by estesp** [comment]
It is strange that I'm just learning this, but a weird mismatch that we actually query the Linux UTS subsystem for the daemon's `architecture` setting, but all image handling is done with `GOOS` and `GOARCH`.  It was confusing to me that an image was trying to match on something that Go doesn't even acknowledge (`armv7l`) which to me is more the root of the problem than the well understood issue with ARM variants.

Seems we need to at some point resolve matching to all use GOOS/GOARCH (with the added `variant` processing that is already available in OCI and also available when building a manifest list) so that we don't have some resolver of a bunch of random distro `uname` choices back to something "understood".

**Comment by thaJeztah** [comment]
I agree with all of the above; we should fix the current situation, because it's a regression from 17.03;

- use "OS", if it's a known value
- use architecture, if it's a known value
- if not, ignore, and use the same behaviour as docker 17.03

> Wouldn't this PR cause ARM images to be scheduled on x86 and other platforms? I suppose it's better than not working at all, but I'm wondering if we can do better.

This is what 17.03 did, so not a regression. Mixed architecture Swarm clusters are probably rare, and if someone currently uses a mixed architecture (or operating system) cluster, they probably use constraints already to influence scheduling.

In the reporter's case https://github.com/docker/swarmkit/issues/2294, things worked fine in the old situation; the service used an image specific to the architecture, and could just be deployed.

**Comment by justincormack** [comment]
@estesp what Go chooses to do does not correspond to the real world
requirements of users necessarily either. This will likely diverge more in
future too, so I don't think it is a great idea just to base things on Go.

**Comment by estesp** [comment]
@justincormack true, the connection to Go is not to take their view of OS/architecture as gospel, but realize that as Go programs, we have a few options for determining our current runtime architecture and OS; either ask the Go runtime, or try to divine it from Linux syscalls/environment data.

FWIW, IMHO, YMMV :) .. Go is a much more stable source of this information than Linux as distros have shown to be quite open to variance and finding their own way when it comes to embedded architectures and `uname -a` responses :)

To make this more concrete, there is an issue (I think it was over in the OCI spec repo) that lists about 60 different variants (strings) of arch-os tuples on ARM platforms running Linux. Go provides us 2: "arm" and "arm64". That leaves us having to find another method (using the proposed variant field in manifest list/OCI indexes) when building images for v6, v7, v8 specific ARM families, but at least it means we aren't trying to harmonize 60 different strings down to "arm" or "arm64" + a variant list of 3 items. Of course we will have to find a way to query the runtime "variant" when arch = {arm,arm64}, which IMO is the only missing piece, but that only applies to manifest list/index-based image resolving, not single manifest images which only provide us the two pieces of info: OS and arch.

If I haven't rambled on too long (don't answer that), another way to look at this bug is that you will never have a stable comparison between runtime.GOARCH and `uname -a`; and images are created using runtime.GOARCH but they are being compared to output from `uname -a`. It was already shown to be a problem in the old days because x86_64 (a Linux-ism) was being hardcoded in images, but runtime.GOARCH in Go is called "amd64" for the same platform.

**Comment by stevvooe** [comment]
@justincormack UTS doesn't exactly help here. Right now, it returns `armv7l` in certain cases, which I cannot find any information about, anywhere. UTS (or gnu triples) seem identify a toolchain, which can be very different for identical architectures.

Go's approach is at least pragmatic: they try to define a real machine target.

If you can identify a standard approach that actually works, then, by all means, suggest one. Laying this on the footsteps of Go's toolchain is fairly dismissive.

**Comment by alexellis** [comment]
I also tested `17.07.0-ce-rc1` this morning on `armv7l` and after upgrading from 17.05 (fully working with stacks and services) - I get 0/1 replicas after deploying a stack.

In addition there are no error messages on service inspect/ls/ps.

**Comment by andrewhsu** [comment]
@alexellis fyi the code changes in this PR was not included in the 17.07.0-ce-rc1 release because it has not been merged yet. Curious to know if you ran your test with a custom compile of the 17.07.0-ce-rc1 codebase with this PR patch applied?

**Comment by alexellis** [comment]
I think Stefan and myself tried the same install script from Eli -

```
curl -fsSL
https://raw.githubusercontent.com/seemethere/docker-install/update_for_170701rc1/install.sh
| sh
```

**Comment by alexellis** [comment]
I think Stefan and myself tried the same install script from Eli @ Docker Inc via the #arm Slack channel-

```
curl -fsSL
https://raw.githubusercontent.com/seemethere/docker-install/update_for_170701rc1/install.sh
| sh
```

```
 Version:      17.07.0-ce-rc1
 API version:  1.31
 Go version:   go1.8.3
 Git commit:   8c4be39
 Built:        Wed Jul 26 21:00:23 2017
 OS/Arch:      linux/arm
```

**Comment by thaJeztah** [comment]
@alexellis what does 

```bash
docker service inspect -f '{{json .Spec.TaskTemplate.Placement}}' <service-id>
```

show for those services?

**Comment by alexellis** [comment]
```
{"Platforms":[{"Architecture":"arm","OS":"linux"}]}
```

**Comment by alexellis** [comment]
@thaJeztah ... it's still broken in the following version
@andrewhsu 

```
Client:
 Version:      17.07.0-ce-rc2
 API version:  1.31
 Go version:   go1.8.3
 Git commit:   36ce605
 Built:        Mon Aug  7 23:53:53 2017
 OS/Arch:      linux/arm

Server:
 Version:      17.07.0-ce-rc2
 API version:  1.31 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   36ce605
 Built:        Mon Aug  7 23:47:08 2017
 OS/Arch:      linux/arm
 Experimental: false
```

**Comment by justincormack** [comment]
@alexellis nothing has changed since https://github.com/moby/moby/pull/34021#issuecomment-318489513 "fyi the code changes in this PR was not included in the 17.07.0-ce-rc1 release because it has not been merged yet. Curious to know if you ran your test with a custom compile of the 17.07.0-ce-rc1 codebase with this PR patch applied?"

Have you tested it? Does it fix the issue?

**Comment by alexellis** [comment]
Do you have a "custom compile" of Docker CE available for testing?

**Comment by alexellis** [comment]
This works.. I tested here along with rallying some of the community to help with the ARM testing.

https://github.com/alexellis/docker-arm/issues/14#issuecomment-323601517

**Comment by vattybear** [comment]
Concur with @alexellis above - have similarly tested the fix today - commented with details on https://github.com/alexellis/docker-arm/issues/17

**Comment by andrewhsu** [comment]
Is this PR a suitable stepping-stone step in the interim?

File: api/server/router/distribution/distribution_routes.go [hunk 1]
```diff
@@ -123,12 +123,19 @@ func (s *distributionRouter) getDistributionInfo(ctx context.Context, w http.Res
 		if err == nil {
 			err := json.Unmarshal(configJSON, &platform)
 			if err == nil && (platform.OS != "" || platform.Architecture != "") {
+				if platform.Architecture == "arm" || platform.Architecture == "Arm" {
```
Comment (stevvooe) [inline api/server/router/distribution/distribution_routes.go]: I was looking at this approach and doing it on the server side will create future problems for this endpoint. Clearing the architecture should be done on the client side, when setting up the constraint. That way, we can fix this in the future and still be able to support this version of docker.

**Comment by nishanttotla** [comment]
Moved changes to client side.

**Comment by stevvooe** [comment]
LGTM

@nishanttotla Thanks!

File: client/service_create.go [hunk 2]
```diff
@@ -98,6 +98,11 @@ func imageDigestAndPlatforms(ctx context.Context, cli DistributionAPIClient, ima
 	if len(distributionInspect.Platforms) > 0 {
 		platforms = make([]swarm.Platform, 0, len(distributionInspect.Platforms))
 		for _, p := range distributionInspect.Platforms {
+			// clear architecture field for arm. This is a temporary fix.
+			arch := p.Architecture
+			if arch == "arm" || arch == "Arm" {
```
Comment (cpuguy83) [inline client/service_create.go]: if `stringsToLower(arch) == "arm"`

File: client/service_create.go [hunk 3]
```diff
@@ -98,6 +98,11 @@ func imageDigestAndPlatforms(ctx context.Context, cli DistributionAPIClient, ima
 	if len(distributionInspect.Platforms) > 0 {
 		platforms = make([]swarm.Platform, 0, len(distributionInspect.Platforms))
 		for _, p := range distributionInspect.Platforms {
+			// clear architecture field for arm. This is a temporary fix.
```
Comment (cpuguy83) [inline client/service_create.go]: Can you add more detail here as to why it's being cleared?

File: client/service_create.go (hunk 2 above)
Comment (cpuguy83) [reply client/service_create.go]: Also should this be `HasPrefix`?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: client/service_create.go (hunk 2 above)
Comment (nishanttotla) [reply client/service_create.go]: In the images I've seen, it's only been "arm". We could use `HasPrefix` to be extra careful but I don't think it's necessary.

**Comment by nishanttotla** [comment]
@cpuguy83 addressed your comments.

**Comment by cpuguy83** [comment]
Lint errors are being dealt with here: https://github.com/moby/moby/pull/34706

**Comment by nishanttotla** [comment]
This commit will also need vendoring into the `docker/cli`.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by andrewhsu** [comment]
Is this one ready to go or are there any other types of verification that is needed before merging?

**Review by stevvooe — APPROVED** [review]

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Pull request merged**